### PR TITLE
Feature/ios push notifications improves

### PR DESCRIPTION
**Pull request by AlexandraMash**
### Description of Change ###

Added possibility to register categories with actions for iOS push notifications library, as well as option to handle badge in foreground push notifications and set it's value manually

### API Changes ###

Added:
 - `INotificationCategoriesProvider` with default implementation `IosNotificationCategoriesProvider`
 - `void SetBadgeNumber(int badgeNumber)` to `IPushNotificationsService`
 - `const string UNNotificationDismissActionIdentifier` to `IosPushNotificationsService`

Changed:
 - `Initialize(bool showForegroundNotificationsInSystem)` => `Initialize(ForegroundNotificationOptions showForegroundNotificationsInSystemOptions)`

### Platforms Affected ### 

- Core (all platforms)
- iOS
- Android

### Behavioral/Visual Changes ###

None

### Before/After Screenshots ### 

Not applicable

### PR Checklist ###

- [x] Targets the correct branch
- [x] Tests are passing (or failures are unrelated)

### CODE REVIEW

**Comment by wcoder** [comment]
Notes about Android notification badges:
- Android 8+ badges (https://developer.android.com/training/notify-user/badges)
- Android bindings library [ShortcutBadger](https://github.com/wcoder/ShortcutBadger) *optional

File: Softeq.XToolkit.PushNotifications/ForegroundNotificationOptions.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    }
+
+    public static class ForegroundNotificationOptionsExtensions
+    {
+        /// <summary>
+        /// Helper method to identify if we should show notifications in foreground for these options
+        /// </summary>
+        /// <param name="foregroundOptions">Value of current foreground options</param>
+        /// <returns>True if notifications should be shown in foreground, False otherwise</returns>
+        public static bool DoShow(this ForegroundNotificationOptions foregroundOptions) =>
```
Comment (wcoder) [inline Softeq.XToolkit.PushNotifications/ForegroundNotificationOptions.cs]: Mey be to rename to `CanShow` or something else?

File: Softeq.XToolkit.PushNotifications.iOS/IosNotificationCategoriesProvider.cs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+using UIKit;
+using UserNotifications;
+
+namespace Softeq.XToolkit.PushNotifications.iOS
+{
+    public class IosNotificationCategoriesProvider : INotificationCategoriesProvider
+    {
+        public IList<UNNotificationCategory> NotificationCategories { get; } = new List<UNNotificationCategory>();
+
+        protected Dictionary<string, IList<UNNotificationAction>> _actionsForCategories = new Dictionary<string, IList<UNNotificationAction>>();
```
Comment (wcoder) [inline Softeq.XToolkit.PushNotifications.iOS/IosNotificationCategoriesProvider.cs]: can be `readonly`

File: Softeq.XToolkit.PushNotifications.iOS/IosNotificationCategoriesProvider.cs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        /// that were previously added for this Category Id. See <see cref="UNNotificationCategory.FromIdentifier"/>
+        /// </summary>
+        /// <param name="categoryId">The unique identifier for the Category. Should not be empty</param>
+        /// <param name="intentIdentifiers">The intent identifier strings that you want to associate with notifications of this type.
+        /// The Intents framework defines constants for each type of intent that you can associate with your notifications</param>
+        /// <param name="options">Additional options for handling notifications of this type</param>
+        protected void AddCategory(string categoryId, string[] intentIdentifiers, UNNotificationCategoryOptions options)
+        {
+            IList<UNNotificationAction> actions;
+            _actionsForCategories.TryGetValue(categoryId, out actions);
```
Comment (wcoder) [inline Softeq.XToolkit.PushNotifications.iOS/IosNotificationCategoriesProvider.cs]: can be simplified `out var`

File: Softeq.XToolkit.PushNotifications.iOS/IosNotificationCategoriesProvider.cs [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        /// <param name="intentIdentifiers">The intent identifier strings that you want to associate with notifications of this type.
+        /// The Intents framework defines constants for each type of intent that you can associate with your notifications</param>
+        /// <param name="hiddenPreviewsBodyPlaceholder">A placeholder string to display when the user has disabled notification previews for the app.
+        /// Include the characters %u in the string to represent the number of notifications with the same thread identifier</param>
+        /// <param name="options">Additional options for handling notifications of this type</param>
+        protected void AddCategory(string categoryId, string[] intentIdentifiers, string hiddenPreviewsBodyPlaceholder,
+            UNNotificationCategoryOptions options)
+        {
+            IList<UNNotificationAction> actions;
+            _actionsForCategories.TryGetValue(categoryId, out actions);
```
Comment (wcoder) [inline Softeq.XToolkit.PushNotifications.iOS/IosNotificationCategoriesProvider.cs]: also `out var`

File: Softeq.XToolkit.PushNotifications/IPushNotificationsService.cs [hunk 5]
```diff
@@ -14,12 +14,11 @@ public interface IPushNotificationsService : IPushNotificationsReceiver
         ///     On iOS call this method before your app finishes launching (in WillFinishLaunching or FinishedLaunching of AppDelegate).
         ///     On Android call this method in OnCreate of Application
         /// </summary>
-        /// <param name="showForegroundNotificationsInSystem">
-        ///     If true, notifications received in foreground will be shown in system as well as all other notifications.
-        ///     Note: On Android this parameter applies only if push notification contains 'notification' part, if it only contains 'data'
-        ///     part notifications will always be shown in system
+        /// <param name="showForegroundNotificationsInSystemOptions">
+        ///     If Show or ShowWithBadge, notifications received in foreground will be shown in system as well as all other notifications.
```
[the remark this answers is not in the document]
Comment (AlexandraMash) [reply Softeq.XToolkit.PushNotifications/IPushNotificationsService.cs]: Added

File: Softeq.XToolkit.PushNotifications/ForegroundNotificationOptions.cs (hunk 1 above)
Comment (AlexandraMash) [reply Softeq.XToolkit.PushNotifications/ForegroundNotificationOptions.cs]: Renamed

File: Softeq.XToolkit.PushNotifications.iOS/IosNotificationCategoriesProvider.cs (hunk 2 above)
Comment (AlexandraMash) [reply Softeq.XToolkit.PushNotifications.iOS/IosNotificationCategoriesProvider.cs]: Fixed

File: Softeq.XToolkit.PushNotifications.iOS/IosNotificationCategoriesProvider.cs (hunk 3 above)
Comment (AlexandraMash) [reply Softeq.XToolkit.PushNotifications.iOS/IosNotificationCategoriesProvider.cs]: Fixed

File: Softeq.XToolkit.PushNotifications.iOS/IosNotificationCategoriesProvider.cs (hunk 4 above)
Comment (AlexandraMash) [reply Softeq.XToolkit.PushNotifications.iOS/IosNotificationCategoriesProvider.cs]: Fixed

File: Softeq.XToolkit.PushNotifications.iOS/IosNotificationCategoriesProvider.cs [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+            if (UIDevice.CurrentDevice.CheckSystemVersion(12, 0))
+            {
+                messageCategory = UNNotificationCategory.FromIdentifier(categoryId,
+                    actions?.ToArray() ?? new UNNotificationAction[] { }, intentIdentifiers,
+                    hiddenPreviewsBodyPlaceholder, new NSString(categorySummaryFormat), options);
+            }
+            else
+            {
+                messageCategory = UNNotificationCategory.FromIdentifier(categoryId,
```
Comment (TimofeyBurak) [inline Softeq.XToolkit.PushNotifications.iOS/IosNotificationCategoriesProvider.cs]: Seems like in this case logic is completely duplicates already existed AddCategory() method (the one without categorySummaryFormat). Can we check system version first and if iOS version is less than 12.0 - just cal that method?

File: Softeq.XToolkit.PushNotifications.iOS/IosNotificationCategoriesProvider.cs [hunk 7]
```diff
@@ -0,0 +1,135 @@
+﻿using System.Collections.Generic;
```
Comment (TimofeyBurak) [inline Softeq.XToolkit.PushNotifications.iOS/IosNotificationCategoriesProvider.cs]: Seems like file header is missing

File: Softeq.XToolkit.PushNotifications.iOS/IosNotificationCategoriesProvider.cs [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+using UIKit;
+using UserNotifications;
+
+namespace Softeq.XToolkit.PushNotifications.iOS
+{
+    public class IosNotificationCategoriesProvider : INotificationCategoriesProvider
+    {
+        public IList<UNNotificationCategory> NotificationCategories { get; } = new List<UNNotificationCategory>();
+
+        protected readonly Dictionary<string, IList<UNNotificationAction>> _actionsForCategories = new Dictionary<string, IList<UNNotificationAction>>();
```
Comment (TimofeyBurak) [inline Softeq.XToolkit.PushNotifications.iOS/IosNotificationCategoriesProvider.cs]: Fields should be defined before properties. Also, does it really have to be protected? Seems like we already define all methods to manipulate this dictionary, so it shouldn't be modified directly

File: Softeq.XToolkit.PushNotifications/ForegroundNotificationOptions.cs [hunk 9]
```diff
@@ -0,0 +1,32 @@
+﻿namespace Softeq.XToolkit.PushNotifications
```
Comment (TimofeyBurak) [inline Softeq.XToolkit.PushNotifications/ForegroundNotificationOptions.cs]: Seems like file header is missing

File: Softeq.XToolkit.PushNotifications.iOS/IosNotificationCategoriesProvider.cs (hunk 6 above)
Comment (AlexandraMash) [reply Softeq.XToolkit.PushNotifications.iOS/IosNotificationCategoriesProvider.cs]: Done

File: Softeq.XToolkit.PushNotifications.iOS/IosNotificationCategoriesProvider.cs (hunk 7 above)
Comment (AlexandraMash) [reply Softeq.XToolkit.PushNotifications.iOS/IosNotificationCategoriesProvider.cs]: Fixed

File: Softeq.XToolkit.PushNotifications.iOS/IosNotificationCategoriesProvider.cs (hunk 8 above)
Comment (AlexandraMash) [reply Softeq.XToolkit.PushNotifications.iOS/IosNotificationCategoriesProvider.cs]: Fixed

File: Softeq.XToolkit.PushNotifications/ForegroundNotificationOptions.cs (hunk 9 above)
Comment (AlexandraMash) [reply Softeq.XToolkit.PushNotifications/ForegroundNotificationOptions.cs]: Fixed

**Pull request merged**